File: browser/js/core/historyStack.js
import constants from '../config/constants.json' with { type: 'json' };

// Line-snapshot history: deep-copy on push/undo/redo, "step 0 → empty lines, step -1" on undo.

// Depth cap shared with core/state/historyStack.hpp's MAX_STEPS (drift-tested in
// tests/history.test.js); cli's max_states and pystencil's _MAX_STATES carry the same 64.
export const MAX_STEPS = constants.LIMITS.historyMax;

export class HistoryStack {
  constructor() {
    this.history = [];
    this.historyStep = -1;
  }

// baseStep: loadImage passes lines.length > 0 ? 0 : -1, restore passes 0.
  reset(lines, baseStep) {
    const step = baseStep !== undefined ? baseStep : (lines.length > 0 ? 0 : -1);
// A negative step means "no current snapshot": keep the history empty so canRedo() stays
// false (a phantom empty snapshot would surface a stray redo after a blank image).
    this.history = step >= 0 ? [this.#clone(lines)] : [];
    this.historyStep = step;
  }

  push(lines) {
    this.historyStep++;
// Truncate in place: a no-op in the common no-redo case, no reallocation per push.
    if (this.history.length > this.historyStep) this.history.length = this.historyStep;
    this.history.push(this.#clone(lines));
// Evict the oldest and shift the cursor down by as many; undoing off the trimmed front
// still ends at the "empty lines, step -1" stop.
    if (this.history.length > MAX_STEPS) {
      this.historyStep -= this.history.splice(0, this.history.length - MAX_STEPS).length;
    }
  }

  canUndo() {
    return this.historyStep >= 0;
  }

  canRedo() {
    return this.historyStep < this.history.length - 1;
  }

// The lines to apply, or null.
  undo() {
    if (this.historyStep > 0) {
      this.historyStep--;
      return this.#clone(this.history[this.historyStep]);
    } else if (this.historyStep === 0) {
      this.historyStep = -1;
      return [];
    }
    return null;
  }

// The lines to apply, or null.
  redo() {
    if (this.historyStep < this.history.length - 1) {
      this.historyStep++;
      return this.#clone(this.history[this.historyStep]);
    }
    return null;
  }

// Stored history must be immune to later mutation of live lines (their points arrays).
  #clone(lines) {
    return structuredClone(lines);
  }
}

File: core/state/historyStack.cpp
#include "historyStack.hpp"

namespace stencil::core {

  HistoryStack::HistoryStack() {
    historyStep_ = -1;
  }

  void HistoryStack::reset(const Lines& lines) {
    reset(lines, lines.empty() ? -1 : 0);
  }

  void HistoryStack::reset(const Lines& lines, int baseStep) {
    history_.clear();
    // A negative base step is "no current snapshot": an empty history keeps canRedo()
    // false (a phantom snapshot would make step -1 < size - 1).
    if (baseStep >= 0) history_.push_back(lines);
    historyStep_ = baseStep;
  }

  void HistoryStack::push(const Lines& lines) {
    ++historyStep_;
    history_.resize(static_cast<std::size_t>(historyStep_));  // drop redo branch
    history_.push_back(lines);
    // Bound the depth: drop the oldest, shift the cursor down as far, so it still names
    // the snapshot just pushed. Undoing off the trimmed front still hits the step -1 stop.
    if (history_.size() > MAX_STEPS) {
      const std::size_t drop = history_.size() - MAX_STEPS;
      history_.erase(history_.begin(), history_.begin() + static_cast<std::ptrdiff_t>(drop));
      historyStep_ -= static_cast<int>(drop);
    }
  }

  bool HistoryStack::canUndo() const {
    return historyStep_ >= 0;
  }

  bool HistoryStack::canRedo() const {
    return historyStep_ < static_cast<int>(history_.size()) - 1;
  }

  std::optional<Lines> HistoryStack::undo() {
    if (historyStep_ > 0) {
      --historyStep_;
      return history_[static_cast<std::size_t>(historyStep_)];
    }
    if (historyStep_ == 0) {
      historyStep_ = -1;
      return Lines{};
    }
    return std::nullopt;
  }

  std::optional<Lines> HistoryStack::redo() {
    if (historyStep_ < static_cast<int>(history_.size()) - 1) {
      ++historyStep_;
      return history_[static_cast<std::size_t>(historyStep_)];
    }
    return std::nullopt;
  }

}
